embed/serve.go

```diff
@@ -325,6 +331,16 @@ func (ac *accessController) ServeHTTP(rw http.ResponseWriter, req *http.Request)
 			http.Error(rw, errCVE20185702(host), 421)
 			return
 		}
+	} else if ac.s.Cfg.ClientCertAuthEnabled {
+		for _, chains := range req.TLS.VerifiedChains {
+			if len(chains) < 1 {
+				continue
+			}
+			if len(chains[0].Subject.CommonName) != 0 {
+				http.Error(rw, "client sending requests against gateway must have empty CommonName in its cert", 400)
+				return
+			}
+		}
```

Comment:
Do we need to stop all requests with error in this case since we have already disabled CN auth for gRPC-gateway to server? Will this also return error for clientv3 CN auth requests?

Reply:
Sorry I forgot to add a condition for checking URL which indicates a request is targeting gateway.

Reply:
While I understand the concern with regards to gRPC-gateway and CN certs I don't know if we should stop what could be valid authentication request because of the common name in the cert which is benign from the client. We document that CN auth is not supported for gRPC-gateway already. I know there is a balance between usability and security here could you give you feelings on this?

Reply:
Do you mean this PR lets etcd return an error to a valid `Authenticate()` RPC? I think replying to the RPC if grpc gateway and cert auth is enabled is meaningless because CN based auth has higher priority than token based auth. The client will never issue RPCs successfully after the `Authenticate()` call. Returning an error ASAP is user friendly.

Reply:
To be clear the calling client to gRPC-gateway that certificate CN is never used for auth. Even before my patch for 3.3.11. As this code blocks HTTP traffic because it has CN I just wonder if this is necessary, thanks.

Reply:
I see. My intention is like this: if client's cert has an empty CN, current error message `{"error":"etcdserver: user name is empty","code":3}` is a correct one from an etcd server with auth and `--client-cert-auth` because it has empty CN actually. If client's cert has non empty CN, the error message is incorrect. Users might be very confused and check their cert (and try with other certs?). Even if the doc mentions about the problem, users will be able to understand why etcd is returning the error immediately.

Also it will be a double check for the security problem. The problem is that requests from the clients can invoke internal requests with a different CN than their own. So if etcd can deny in `accessController` layer, it will be good for defensive purpose.

How do you think?

Reply:
I would personally would like to add your option for enabling/disabling gRPC-gateway and drop this change. Then we can look at additional changes in another PR.

Reply:
Do you mean I should open a new PR for the second and third commits?

Reply:
@gyuho can I also have your opinion about the error message and improving defensiveness?

Reply:
@mitake @hexfusion Yeah, the error message is clear about ignoring the common name for auth.

lgtm